### PR TITLE
Automate the Rust Code Generation

### LINKED ISSUE #942: Automate the Rust Code Generation
### Summary
Now that the flexer can generate rust code for the DFA, it is important that we find a way to automate this generation from the lexer definitions to ensure that things stay in sync with the definition.

### Value
We won't get out of sync as might have otherwise happened if it was done by hand. 

### Specification

- [ ] Work out how to automate the generation of the lexer code at build time. You can assume that a lexer definition will construct rust code in an appropriate format (a string, for example). 
- [ ] Document this in the flexer design docs.
- [ ] Implement this automation step for use by the Enso lexer.
- [ ] This may involve `build.rs` work or procedural macros.
- [ ] It must recompile the lexer if the lexer definition changes _or_ if the definition of flexer changes, but should not do it spuriously (i.e. unconditionally recompiling is _not_ an option).

### Acceptance Criteria & Test Cases

- The lexer code is automatically generated at build time by cargo.

### PR DESCRIPTION
**Pull request by Kesanov**
### Pull Request Description
 
Closes #942.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: lib/rust/flexer/Cargo.toml [hunk 1]
```diff
@@ -3,7 +3,7 @@ name    = "flexer"
 version = "0.1.0"
 authors = [
   "Enso Team <enso-dev@enso.org>",
-  "Ara Adkins <ara.adkins@enso.org"
+  "Ara Adkins <ara.adkins@enso.org>"
```
Comment (wdanilo) [inline lib/rust/flexer/Cargo.toml]: Ara is part of Enso team, so there is no need to keep that separate. We are not doing it anywhere else. Lets reovethis entrty

File: lib/rust/parser/definition/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1 @@
+pub mod parser;
```
Comment (wdanilo) [inline lib/rust/parser/definition/src/lib.rs]: nl - beforerequesting review, please make sure your code is listed and also make sure you do self-review first - just browse the code on github

File: docs/parser/flexer.md [hunk 3]
```diff
@@ -88,6 +88,32 @@ they mostly choose the _longest_ match instead. Once the pattern is matched, the
 associated code is executed and the process starts over again until the input
 stream has been consumed.
 
+### Automatic code generation
+
+In order to avoid getting lexer definition out of sync with its implementation
```
Comment (iamrecursion) [inline docs/parser/flexer.md]:
```suggestion
### Automatic Code Generation
In order to avoid getting lexer definition out of sync with its implementation
```

You also need to update the table of contents to reflect the new section that you've added.

File: docs/parser/flexer.md [hunk 4]
```diff
@@ -88,6 +88,32 @@ they mostly choose the _longest_ match instead. Once the pattern is matched, the
...
+(generated engine), it is required to create a separate crate for the generated 
+engine with lexer definition as one of its dependencies.
+
+This separation enables a call to `specialize` (function that generates the engine)
+in `build.rs` during compilation. The output can be stored in new file i.e. 
+`lexer-engine.rs` and exported with `include!("lexer-engine.src")`. The project
```
Comment (iamrecursion) [inline docs/parser/flexer.md]:
```suggestion
`lexer-engine.rs` and exported with `include!("lexer-engine.rs")`. The project
```

File: docs/parser/flexer.md [hunk 5]
```diff
@@ -88,6 +88,32 @@ they mostly choose the _longest_ match instead. Once the pattern is matched, the
...
+    - src/
+      - lexer.rs
+    - cargo.toml
+
+  - lexer-engine/
+    - src/
+      - lexer.rs <-- include!("lexer-engine.rs")
+    - build.rs   <-- calls `lexer-definition::Lexer::new()::specialize()`
+                  -- and saves its output to `src/lexer-engine.rs`
+    - cargo.toml <-- lexer-definition is in dependencies and build-dependencies
```
Comment (iamrecursion) [inline docs/parser/flexer.md]:
```suggestion
- lib/rust/lexer/
  - definition/
    - src/
      - lexer.rs
    - cargo.toml

  - generation/
    - src/
      - lexer.rs <-- include!("lexer-engine.rs")
    - build.rs   <-- calls `lexer-definition::Lexer::new()::specialize()`
                  -- and saves its output to `src/lexer-engine.rs`
    - cargo.toml <-- lexer-definition is in dependencies and build-dependencies
```

File: lib/rust/parser/engine/build.rs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+use std::fs::File;
+use std::io::prelude::*;
+use parser_definition::lexer::Lexer;
+use flexer::Flexer;
+
+
+/// Generates lexer engine and saves the result into a file `src/lexer-engine.rs`.
+/// The content of the generated file can be used trough the `include!` macro.
```
Comment (iamrecursion) [inline lib/rust/parser/engine/build.rs]:
```suggestion
/// Generates the lexer engine and saves the result into the file `src/lexer-engine.rs`.
/// 
/// The content of the generated file can be used through the `include!` macro.
```

Please use spell check! IntelliJ has it built in.

File: lib/rust/parser/engine/src/lexer.rs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#![allow(clippy::all)]
+#![allow(unknown_lints)]
+#![allow(missing_copy_implementations)]
+#![allow(missing_debug_implementations)]
+#![allow(missing_docs)]
+#![allow(trivial_casts)]
+#![allow(trivial_numeric_casts)]
+#![allow(unsafe_code)]
+#![allow(unused_import_braces)]
```
Comment (iamrecursion) [inline lib/rust/parser/engine/src/lexer.rs]: Why are these here?

File: lib/rust/parser/engine/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,19 @@
+#![feature(test)]
+#![deny(unconditional_recursion)]
+#![warn(missing_copy_implementations)]
+#![warn(missing_debug_implementations)]
+#![warn(missing_docs)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unsafe_code)]
+#![warn(unused_import_braces)]
```
Comment (iamrecursion) [inline lib/rust/parser/engine/src/lib.rs]: Why?

File: lib/rust/parser/engine/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,6 @@
+use parser_engine::lexer::Lexer;
```
Comment (iamrecursion) [inline lib/rust/parser/engine/src/main.rs]: Why does this file exist?

File: lib/rust/parser/engine/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+//! This library exposes the specialized version of enso lexer.
+//! Its sole purpose is to avoid getting lexer definition out of sync with its implementation
+//! (the generated engine), which requires the engine to live in separate crate.
+//!
+//! This separation enables a call to `lexer.specialize` (function that generates the lexer engine)
+//! inside `build.rs` during compilation. Its output is then stored in a new file `lexer-engine.rs`
+//! and exported by `lexer.rs`.
```
Comment (iamrecursion) [inline lib/rust/parser/engine/src/lib.rs]:
```suggestion
//! This library exposes the specialized version of the enso lexer.
//!
//! Its sole purpose is to avoid the lexer definition getting out of sync with its implementation
//! (the generated engine), which requires the engine to live in a separate crate.
//!
//! This separation enables a call to `lexer.specialize` (the function that generates the lexer engine)
//! inside `build.rs` during compilation. Its output is then stored in a new file `lexer-engine.rs`
//! and exported by `lexer.rs`.
```

Please remember, doc comments should have a short, single-sentence header, then paragraphs explaining things. Paragraphs require a blank line as this is markdown.

The note about `lexer.specialize` should have a TODO comment noting that this should link once that function exists.

**Review by iamrecursion — CHANGES_REQUESTED** [review]
Everywhere I have made suggestions, please ensure that they match the wrapping requirements. Please also add a test of the functionality (you can adapt the `main` file).

File: lib/rust/parser/engine/src/lexer.rs (hunk 7 above)
Comment (Kesanov) [reply lib/rust/parser/engine/src/lexer.rs]: Because this file contains the generated lexer, which I assume our code guidelines should not care about.

File: lib/rust/parser/engine/src/lib.rs (hunk 8 above)
Comment (Kesanov) [reply lib/rust/parser/engine/src/lib.rs]: Don't we use these warnings in every crate?

File: lib/rust/parser/engine/src/main.rs (hunk 9 above)
Comment (Kesanov) [reply lib/rust/parser/engine/src/main.rs]: It should be replaced with a proper test.

File: lib/rust/parser/engine/src/lib.rs (hunk 8 above)
Comment (iamrecursion) [reply lib/rust/parser/engine/src/lib.rs]: Oh, sorry. I misread this as `allow`.

File: lib/rust/parser/engine/src/lexer.rs (hunk 7 above)
Comment (iamrecursion) [reply lib/rust/parser/engine/src/lexer.rs]: So these should be put in the _generated_ code, not in the file including it.

File: lib/rust/parser/engine/src/main.rs (hunk 9 above)
Comment (iamrecursion) [reply lib/rust/parser/engine/src/main.rs]: So please make this a test, as I mentioned in the overall review comment.

File: docs/parser/flexer.md [hunk 11]
```diff
@@ -88,6 +89,38 @@ they mostly choose the _longest_ match instead. Once the pattern is matched, the
 associated code is executed and the process starts over again until the input
 stream has been consumed.
 
+### Automated code generation
```
Comment (iamrecursion) [inline docs/parser/flexer.md]:
```suggestion
### Automated Code Generation
```
Titles should use title case.

File: docs/parser/flexer.md [hunk 12]
```diff
@@ -88,6 +89,38 @@ they mostly choose the _longest_ match instead. Once the pattern is matched, the
...
+In order to avoid getting lexer definition out of sync with its implementation
+(generated engine), it is required to create a separate crate for the generated 
+engine with lexer definition as one of its dependencies.
```
Comment (iamrecursion) [inline docs/parser/flexer.md]:
```suggestion
In order to avoid the lexer definition getting out of sync with its implementation
(the generated engine), it is necessary to create a separate crate for the generated 
engine that has the lexer definition as one of its dependencies.
```

File: docs/parser/flexer.md [hunk 13]
```diff
@@ -88,6 +89,38 @@ they mostly choose the _longest_ match instead. Once the pattern is matched, the
...
+
+This separation enables a call to `flexer.generate_specialized_code()` in `build.rs` 
+(or macro) during compilation. The output can be stored in new file i.e. 
+`lexer-engine.rs` and exported with `include!("lexer-engine.rs")`. The project
+structure should therefore look like:
```
Comment (iamrecursion) [inline docs/parser/flexer.md]:
```suggestion
This separation enables a call to `flexer.generate_specialized_code()` in `build.rs` 
(or a macro) during compilation. The output can be stored in a new file i.e. 
`lexer-engine.rs` and exported from the library with `include!("lexer-engine.rs")`. The project
structure therefore appears as follows:
```

File: docs/parser/flexer.md [hunk 14]
```diff
@@ -88,6 +89,38 @@ they mostly choose the _longest_ match instead. Once the pattern is matched, the
...
+    - build.rs   <-- calls `lexer_definition::lexer_source_code()`
+                  -- and saves its output to `src/lexer-engine.rs`
+    - cargo.toml <-- lexer-definition is in dependencies and build-dependencies
+```
+
+With this design `flexer.generate_specialized_code()` is going to be executed on each rebuild
+of `lexer/generation`. Therefore, `generation` should contain only minimum amount of logic 
+(i.e. tests should be in separate crate) and its dependencies should optimally involve only 
+such code which directly influences the content of generated code (in order to minimize 
+the unnecessary calls to expensive flexer specialization).
```
Comment (iamrecursion) [inline docs/parser/flexer.md]:
```suggestion
With this design, `flexer.generate_specialized_code()` is going to be executed on each rebuild
of `lexer/generation`. Therefore, `generation` should contain only the minimum amount of logic 
(i.e. tests should be in separate crate) and its dependencies should optimally involve only 
such code which directly influences the content of generated code (in order to minimize 
the unnecessary calls to expensive flexer specialization).
```

File: lib/rust/lexer/generation/src/lexer.rs [hunk 15]
```diff
@@ -0,0 +1,4 @@
+//! This file includes the source code generated by lexer specialization.
+
+
```
Comment (iamrecursion) [inline lib/rust/lexer/generation/src/lexer.rs]:
```suggestion
```

File: lib/rust/lexer/tests/src/main.rs [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+#![warn(missing_debug_implementations)]
+#![warn(missing_docs)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unsafe_code)]
+#![warn(unused_import_braces)]
+
+//! This file tests the generated source code of the enso lexer.
+
+
```
Comment (iamrecursion) [inline lib/rust/lexer/tests/src/main.rs:13]: Missing heading.

File: lib/rust/lexer/tests/src/main.rs [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+fn main() {
+    println!("This needs to be here because the crate isn't a library.")
+}
+
+#[cfg(test)]
+mod tests {
+    use lexer_generation::lexer::Lexer;
+
```
Comment (iamrecursion) [inline lib/rust/lexer/tests/src/main.rs:26]: Wrong number of blank lines.

**Review by iamrecursion — CHANGES_REQUESTED** [review]
Please make sure the suggestions are applied but wrapped to the 80-character limit.

File: docs/parser/flexer.md [hunk 18]
```diff
@@ -88,6 +89,41 @@ they mostly choose the _longest_ match instead. Once the pattern is matched, the
...
+```
+
+With this design, `flexer.generate_specialized_code()` is going to be executed 
+on each rebuild of `lexer/generation`. Therefore, `generation` should contain
+only the minimum amount of logic (i.e. tests should be in separate crate) and
+its dependencies should optimally involve only such code which directly 
+influences the content of generated code (in order to minimize the
+unnecessary calls to expensive flexer specialization).
+
+
```
Comment (iamrecursion) [inline docs/parser/flexer.md]:
```suggestion
```

File: lib/rust/flexer/src/lib.rs [hunk 19]
```diff
@@ -19,3 +19,17 @@ pub mod group;
 pub mod data;
 
 pub use enso_prelude as prelude;
+
+
+/// The flexer is an engine for generating lexers. Akin to flex and other lexer generators,
+/// it is given a definition as a series of rules from which it then generates code for a highly
+/// optimised lexer implemented on top of [DFA](https://en.wikipedia.org/wiki/Deterministic_finite_automaton).
```
Comment (iamrecursion) [inline lib/rust/flexer/src/lib.rs]:
```suggestion
/// optimised lexer implemented on top of a [DFA](https://en.wikipedia.org/wiki/Deterministic_finite_automaton).
```

File: lib/rust/flexer/src/lib.rs [hunk 20]
```diff
@@ -19,3 +19,17 @@ pub mod group;
...
+pub trait Flexer {
+    /// Creates a new lexer.
+    fn new() -> Self;
+
+    /// Returns a code for a highly-optimised lexer implemented on top of finite-state-automata.
```
Comment (iamrecursion) [inline lib/rust/flexer/src/lib.rs]:
```suggestion
    /// Returns a code for a highly-optimised lexer implemented on top of a finite-state-automaton.
```

File: lib/rust/lexer/definition/src/lexer.rs [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+
+use flexer::Flexer;
+
+
+
+// ========================
+// === Lexer Definition ===
+// ========================
+
+/// The definition of enso lexer that is responsible for lexing the enso source code.
```
Comment (iamrecursion) [inline lib/rust/lexer/definition/src/lexer.rs:11]:
```suggestion
/// The definition of the Enso lexer.
```

File: lib/rust/lexer/generation/build.rs [hunk 22]
```diff
@@ -0,0 +1,20 @@
+use std::fs::File;
+use std::io::prelude::*;
+use lexer_definition::lexer::Lexer;
+use flexer::Flexer;
+
+
+
+/// Generates the lexer engine and saves the result into a file `src/lexer-engine.rs`.
```
Comment (iamrecursion) [inline lib/rust/lexer/generation/build.rs]:
```suggestion
/// Generates the lexer engine and saves the result into the file `src/lexer-engine.rs`.
```

File: lib/rust/lexer/generation/src/lib.rs [hunk 23]
```diff
@@ -0,0 +1,9 @@
+//! This library exposes the specialized version of the enso lexer.
```
Comment (iamrecursion) [inline lib/rust/lexer/generation/src/lib.rs]:
```suggestion
//! This library exposes the specialized version of the Enso lexer.
```

**Review by iamrecursion — CHANGES_REQUESTED** [review]

File: tools/run [hunk 24]
```diff
@@ -70,14 +59,16 @@ async function withCwd(dir, fn) {
...
-    dirs = fss.readdirSync(sourceDir, {withFileTypes: true})
-        .filter(f => f.isDirectory())
-        .map(d => path.join(sourceDir, d.name))
+    content = fss.readFileSync(path.join(rootDir, 'Cargo.toml'), {encoding: 'utf-8'})
+    content = content.replace(/\s+|"/g, '')
+    members = content.split('members')[1].split('[')[1].split(']')[0]
+    dirs    = members.split(',').map(d => path.join(rootDir, d))
 
-    dirs.forEach(d => withCwd(d, testLibrary))
+    dirs.forEach(d => {
```
Comment (iamrecursion) [inline tools/run]: This also _doesn't_ test `enso-macro-utils`, which should be tested. It's not a proc macro crate. The condition you want is that the folder you enter contains both a `macros` and an `impl` folder, and you only want to test in the `impl` folder.

**Review by iamrecursion — CHANGES_REQUESTED** [review]
Please fix the test script.

File: .gitignore [hunk 25]
```diff
@@ -21,6 +21,7 @@ target/
 Cargo.lock
 **/*.rs.bk
 wasm-pack.log
+lexer-engine.rs
```
Comment (iamrecursion) [inline .gitignore]: This should be ignored at the specific path, not globally.

**Review by iamrecursion — APPROVED** [review]
Good stuff! Please fix the one thing in the `.gitignore` before merging.
